File: WordBreak/WordBreak.js
/*
Given a string s and a dictionary of strings wordDict, return true if s can be segmented into a space-separated sequence of one or more dictionary words.

Note that the same word in the dictionary may be reused multiple times in the segmentation.



Example 1:

Input: s = "leetcode", wordDict = ["leet","code"]
Output: true
Explanation: Return true because "leetcode" can be segmented as "leet code".
Example 2:

Input: s = "applepenapple", wordDict = ["apple","pen"]
Output: true
Explanation: Return true because "applepenapple" can be segmented as "apple pen apple".
Note that you are allowed to reuse a dictionary word.
Example 3:

Input: s = "catsandog", wordDict = ["cats","dog","sand","and","cat"]
Output: false


Constraints:

1 <= s.length <= 300
1 <= wordDict.length <= 1000
1 <= wordDict[i].length <= 20
s and wordDict[i] consist of only lowercase English letters.
All the strings of wordDict are unique.
*/

var wordBreak = function(s, wordDict) {
  const memo = new Array(s.length).fill(-1)
  function dp(i) {
      if (i > s.length - 1) {
          return true
      }
      if (memo[i] !== -1) {
          return memo[i]
      }

      for (let index = 0; index < wordDict.length; index++) {
          const word = wordDict[index];
          if (s.substring(i, word.length + i) === word && dp(word.length + i)) {
              memo[i] = true;
              return true
          }
      }
      memo[i] = false;
      return false
  }

  return dp(0);

};

File: JavaSolution/src/WordBreak.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class WordBreak {
    public boolean check(String s, List<String> wordDict) {
        int[] memo = new int[s.length()];
        Arrays.fill(memo, -1);
        return dfs(0, s, wordDict, memo);
    }
    public boolean dfs(int i, String s, List<String> wordDict, int[] memo) {
        if (i > s.length()) {
            return false;
        }
        if (i == s.length()) {
            return true;
        }
        if (memo[i] != -1) {
            return memo[i] == 1;
        }
        for (String word: wordDict) {
            int len = word.length();
            if (i + len <= s.length() && s.substring(i, i + len).equals(word) && dfs(i + len, s, wordDict, memo)) {
                return true;
            }
        }
        memo[i] = 0;
        return false;
    }
    public boolean checkIterative(String s, List<String> wordDict) {
        boolean[] dp = new boolean[s.length()];
        for (int i = 0; i < s.length(); i++) {
            for (String word: wordDict) {
                if (i - word.length() + 1 < 0) {
                    continue;
                }
                dp[i] = (i - word.length() == -1 || dp[i - word.length()]) && s.substring(i - word.length() + 1, i + 1).equals(word);
                if (dp[i]) {
                    break;
                }
            }
        }
        return dp[s.length() - 1];
    }
    public static void main(String[] args) {
        WordBreak wordBreak = new WordBreak();
        String s = "leetcode";
        List<String> wordDict = Arrays.asList("leet", "code");
        System.out.println(wordBreak.checkIterative(s, wordDict));
    }
}
